### PR TITLE
`Development`: Refactor competencies management page to signals

### PR DESCRIPTION
**Pull request by JohannesWt**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

According to the new client coding guidelines Artemis should move to using signals step by step. This PR refactors the competency management page to using signals.

### Description

This PR changes the current competency management page to use signals. Visually nothing changed.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to course competency management 
3. verify that a competency and prerequisite can be created and that they are shown in the table
4. verify that a competency/prerequiste can be deleted
5. verify that the generate competencies button is visible (on TS3 as Iris is required)

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-management.component.ts | 92.04% | ✅ ❌ |

### Screenshots

unchanged

## Summary by CodeRabbit

- **New Features**
	- Enhanced reactivity in the Competency Management component, improving data handling and state management.
	- Updated loading indicators and competency retrieval methods for better user experience.
	- Improved structure for handling competencies in the Competency Management Table component.

- **Bug Fixes**
	- Improved error handling and alert notifications for data fetching issues.

- **Tests**
	- Refined test suite for better coverage and reliability, including new tests for loading states and error handling.

### CODE REVIEW

**Review by Cathy0123456789 — DISMISSED** [review]
Tested on TS3. Creating, editing and deleting competencies / prerequisites still works, as well as linking and unlinking lecture units. Also the "Generate Competencies" button is shown.

**Review by flbrgit — DISMISSED** [review]
Tested on TS3. Functionality works as described and the button to generate competencies is also shown. Linking to units, deleting and creating as well as editing works without noticable bugs

**Review by ItsaaaMeMario — APPROVED** [review]
I tested on TS3 with artemis_test_user_20, I was flawlessly able to create/delete/edit prerequisites/competencies. The "generate Competencies" button is also visible.

File: src/test/javascript/spec/component/competencies/competency-management/competency-management.component.spec.ts [hunk 1]
```diff
@@ -60,6 +65,18 @@ describe('CompetencyManagementComponent', () => {
...
+                    useClass: MockAlertService,
+                },
+                {
+                    provide: ProfileService,
+                    useClass: MockProfileService,
+                },
+                {
+                    provide: AlertService,
+                    useClass: MockAlertService,
+                },
```
Comment (JohannesStoehr) [inline src/test/javascript/spec/component/competencies/competency-management/competency-management.component.spec.ts:79]:
```suggestion
                {
                    provide: AlertService,
                    useClass: MockAlertService,
                },
                {
                    provide: ProfileService,
                    useClass: MockProfileService,
                },
```
One should be enough

**Review by JohannesStoehr — APPROVED** [review]
Code looks good apart from one small issue:

**Review by HawKhiem — APPROVED** [review]
Competencies and prerequesites can be added and deleted as described. The generate competencies button is also visible. But when I click on it it takes really long to load

image

**Review by zagemello — DISMISSED** [review]
image
Works as expected, tested on TS3

**Review by SindiBuklaji — DISMISSED** [review]
Tested on TS3. Everything works as expected.👍

**Review by EneaGore — APPROVED** [review]
code

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by JohannesStoehr — APPROVED** [review]

**Review by ItsaaaMeMario — DISMISSED** [review]
Tested on TS3, everything works as expected.

**Review by sachmii — APPROVED** [review]
Tested on TS1 and TS3, works fine (generate competency on TS3).

**Review by kevinfischer4 — APPROVED** [review]
Tested on TS3, creating competencies and prerequisites works as expected. The "Generate Competencies"-button is also visible.

**Review by JerroyTan — COMMENTED** [review]
[Tested on TS1]
Did not get the chance to test on TS3 as it was locked the whole time I was doing my reviews.
All basic functions (except for the last one which required TS3) were working as expected.

Issue: Import competencies only works for single imports. When using "Import all competencies of a course", I select a course (which I know has competencies as I can import them single) and it says page does not exist.
image

**Review by sawys777 — COMMENTED** [review]
Tested on TS3, no problem in adding and deleting competencies and the generate competencies was visible as well.

**Review by Haoyuli2002 — COMMENTED** [review]
Tested on TS 3 as User 17(Instructor). I could successfully add the competency and prerequisite. Also I could delete them afterwards and the generation button is visible. Everything is working as expected and working well.

**Review by zagemello — DISMISSED** [review]
All works as expected, tested on TS4 (was not able to access TS3)

**Review by az108 — APPROVED** [review]
Code 👍

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Merge conflict

**Review by az108 — APPROVED** [review]
reapprove merge

**Review by sachmii — APPROVED** [review]
reapprove

**Review by JohannesStoehr — APPROVED** [review]
Reapprove

**Review by EneaGore — APPROVED** [review]
reapprove

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**